Add have rich text matcher

VSPPedro opened the pull request:
Hi, this PR solves the issue #1252. 

It's a draft for now, I would love to know your opinion about it.

One of the things that needs to be improved in this PR is the creation of the `action_text_rich_texts` table. I'm out of ideas for now, any suggestions?

mcmire commented:
@VSPPedro How does `action_text_rich_texts` usually get generated in a Rails app? Is it using a generator, or does it get generated automatically on the fly? (Sorry, I've never used ActionText so I don't really know how it works.)

VSPPedro commented:
Hi, @mcmire, to make the action text to work in a Rails app it's necessary to run: `rails action_text:install`. This command it will install the yarn package necessary and add a migration with the `action_text_rich_texts` table:
```
# This migration comes from action_text (originally 20180528164100)
class CreateActionTextTables < ActiveRecord::Migration[6.0]
  def change
    create_table :action_text_rich_texts do |t|
      t.string     :name, null: false
      t.text       :body, size: :long
      t.references :record, null: false, polymorphic: true, index: false

      t.timestamps

      t.index [ :record_type, :record_id, :name ], name: "index_action_text_rich_texts_uniqueness", unique: true
    end
  end
end
```

guialbuk commented on lib/shoulda/matchers/active_record/have_rich_text_matcher.rb:
```diff
@@ -0,0 +1,77 @@
+        end
+
+        def has_expected_action_text?
+          @subject.send(rich_text_attribute).class == ActionText::RichText
```
This can raise `uninitialized constant` if `ActionText::RichText` is not available for the Rails version or available but not loaded.
A solution would be to compare `class.name` with a `'ActionText::RichText'` string

VSPPedro replied:
True!

I'll update it! Thank you!

VSPPedro replied:
Done!

guialbuk commented on spec/support/unit/helpers/active_record_versions.rb:
```diff
@@ -50,5 +50,9 @@ def active_record_supports_expression_indexes?
     end
+
+    def active_record_supports_has_rich_text?
+      active_record_version >= '6.0.0.beta1'
```
It's possible to have Rails 6 and not load Action Text:

https://github.com/rails/rails/blob/34d66ca6bab109540299a7b30980627bea37e323/railties/lib/rails/all.rb#L16

So to check if Action Text is available, you can check for `ActionText::RichText` or `ActionText::Engine`.

VSPPedro replied:
You're right!

I'll add this validation!

VSPPedro replied:
I updated to this:
```
    def active_record_supports_has_rich_text?
      active_record_version >= '6.0.0.beta1' &&
        defined?(ActionText::RichText)
    end
```
Was that what you had in mind?

guialbuk replied:
I don't think you need to check the rails version (some people are even backporting ActionText to 5.2).
Just `defined?(ActionText::RichText)`should be enough 👍

guialbuk commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,88 @@
+
+describe Shoulda::Matchers::ActiveRecord::HaveRichText, type: :model do
+
+  if active_record_supports_has_rich_text?
```
It would be nice to have tests for falsy `active_record_supports_has_rich_text?`.

VSPPedro replied:
Sure! Before this branch is merged I'll added that!

VSPPedro replied:
Done!

Please, take a look into the if conditions. The difference between the second and the third is the name of the class inside the message.  

In Rails 4 the test returns a NillClass, but in Rails 5 or greater the test return a String class.

This is what I got when I run the tests with Rails 4: 

``` 
activerecord-4.2.11.1/lib/active_record/connection_adapters/abstract_adapter.rb:84: warning: deprecated Object#=~ is called on Integer; it always returns nil
```

I'll create some `shared examples` for the duplications.

mcmire commented:
@VSPPedro So instead of adding a migration in the test itself, what I would do instead is to actually run the generator you mentioned. To explain how you would do this, before you run unit tests, a full Rails application is actually generated in the background. You can see that happening here: https://github.com/thoughtbot/shoulda-matchers/blob/master/spec/support/unit/load_environment.rb and here is the main method that generates the app: https://github.com/thoughtbot/shoulda-matchers/blob/master/spec/support/unit/rails_application.rb#L74.

Given this, if you add a new method to UnitTests::RailsApplication#generate that does something like this:

``` ruby
if rails_version >= 6 && bundle.includes?('actiontext')
  fs.within_project do
    run_command! `rails action_text:install`
  end
end
```

and then call it in #generate, then that may solve the issue for you. Migrations get run in the #load method after the #create method is called, so you shouldn't have to take care of that step. 

The benefit of this approach is that if Rails for some reason changes ActionText to use a different table, we shouldn't have to update shoulda-matchers in the future to accommodate this — everything should just continue to work. 

Let me know if you have any more questions on this.

VSPPedro commented:
First of all, thanks for all the knowledge you two are sharing with me, I am learning a lot.

@mcmire , I tried to use this:

```
if rails_version >= 6 && bundle.includes?('actiontext')
  fs.within_project do
    run_command! `rails action_text:install`
  end
end
```

 inside the #generate, but without success. These errors were popping out:

```ruby
RAILS_ENV=development environment is not defined in config/webpacker.yml, falling back to production environment
  rake aborted!
  ActiveRecord::AdapterNotSpecified: The `development` database is not configured for the `development` environment.
```

 Only worked when I tried to call it inside the `load` method right after the `load_environment` and with this it worked like a charm.

Another problems that I faced was that the command to install action_text only works with webpacker installed. I saw that the command to create the rails project skip the webpacker installation so I look for other options and I founded the possibility to copy only the migration file with this command: `action_text:install:migrations`. With this it's not more necessary copy paste the migration file inside the tests. I'm kinda happy with this result! Thanks a lot! :+1:

Another thing, I tried to use this method, `bundle.includes?('actiontext')`, but only returned false. It returned true when I added the actiontext as gem in the gemfile, but I don't think this is a good thing to do.

I replaced with: `Gem.loaded_specs.has_key?('actiontext')`. What do you think? Or should I give another try with  `bundle.includes?('actiontext')`? Or should I add the `actiontext` to the gemfile?

mcmire commented:
@VSPPedro Thanks for looking into this further. I'm glad that you were able to get the migrations created! With regard to `bundle.includes?('actiontext')`, I'd forgotten that ActionText is an optional dependency, so that makes that that would fail. Replacing that with `Gem.loaded_specs.has_key?('actiontext')` is fine, although the `bundle` object can ask the question from a Bundler perspective, so I'd prefer we use that so that we don't get false positives. I think you would have to add ActionText to some kind of Gemfile to get this to work, and that's okay. There are two steps to this:

* Instead of the root Gemfile, you'd probably want to add the gem to the Appraisals file, which is sort of a mesh of Gemfiles we maintain for different Rails versions. If you add it to [this spot here](https://github.com/thoughtbot/shoulda-matchers/blob/4369abdd6ff1d2d21a5bc00960323ec1250c16cf/Appraisals#L133), you can run `appraisal install` to install it locally. This will update `gemfiles/rails_6_0.gemfile` and `gemfiles/rails_6_0.gemfile.lock` but that is perfectly okay!
* You may have to ensure that it shows up in the generated application's Gemfile as well. This may not be completely necessary, but if not, then I think it's at least helpful for consistency in case you need to debug this stuff later. You can do that by adding

    ``` ruby
    if rails_version >= 6
      bundle.add_gem 'actiontext'
    end
    ```

    after [this line](https://github.com/thoughtbot/shoulda-matchers/blob/4369abdd6ff1d2d21a5bc00960323ec1250c16cf/spec/support/unit/rails_application.rb#L197).

VSPPedro commented:
It helped a lot, @mcmire! Thanks!

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,137 @@
+        expected_message = 'Expected Post to have :invalid_attribute as a ' \
+          'ActionText::RichText but :invalid_attribute does not exist'
+    
+        aggregate_failures do
```
Instead of using `aggregate_failures` and having to repeat yourself twice, there's a new-ish matcher you can use called `match_against` which checks that a matcher matches when used in the negative and does not match when used in the positive. You can see that in use here: https://github.com/thoughtbot/shoulda-matchers/blob/0011a3bd9a6b07f8234954d0418a74bdeb393db7/spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb#L529-L536. So in this case you would say:

``` ruby
expect { have_rich_text(:invalid_attribute) }.
  not_to match_against(invalid_record).
  and_fail_with(expected_message)
```

By the way, if you want to thoroughly test both versions of the failure message (`failure_message` and `failure_message_when_negated`), you can also do the same thing with the positive cases, so you could say:

``` ruby
expect { have_rich_text(:content) }.
  to match_against(valid_record).
  or_fail_with(expected_message)
```

and this will test that the matcher matches when used in the positive and does not match when used in the negative.

VSPPedro replied:
Much better! Thanks!

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,137 @@
+      )
+    end
+
+    it 'does not matches when the provided attribute it\'s not a ' \
```
What if you used double quotes here? I feel like backslashing the apostrophe makes it less readable.

VSPPedro replied:
True. Done.

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,137 @@
+    end
+  end
+
+  def create_table_posts(options = {})
```
What do you think about using keyword arguments to shorten this a bit, e.g. `def create_table_posts(has_content_attribute: false)`? Same for `new_post`.

VSPPedro replied:
I like it your suggestion. Is more precise then `options = {}`. Thanks!

mcmire commented on lib/shoulda/matchers/active_record/have_rich_text_matcher.rb:
```diff
@@ -0,0 +1,77 @@
+
+        def matches?(subject)
+          @subject = subject
+          @error = run_checks
```
I like how you've done this — that's a good pattern we should use for future matchers.

mcmire commented on lib/shoulda/matchers/active_record/have_rich_text_matcher.rb:
```diff
@@ -0,0 +1,77 @@
+        end
+
+        def failure_message
+          "Expected #{@subject.class} to #{description} but #{@error}"
```
What do you think about making private `attr_reader`s for these instance variables? And also making the `attr_reader` for `@rich_text_attribute` private?

VSPPedro replied:
I don't see a problem. I'll add it now. Thanks!

VSPPedro commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,137 @@
+    end
+
+    context 'a has_rich_text association' do
+      it 'matches when the subject configures has_rich_text for a custom ' \
```
No problem! Done!

VSPPedro commented:
Next thing to do: add [shared_examples](https://relishapp.com/rspec/rspec-core/v/2-0/docs/example-groups/shared-example-group).

My task for tomorrow.

VSPPedro commented:
I removed the WIP, I think is almost done. What do you think?

Tomorrow I'll look for the possibility of adding [shared_examples](https://relishapp.com/rspec/rspec-core/v/2-0/docs/example-groups/shared-example-group).

VSPPedro commented:
Done!

mcmire commented on lib/shoulda/matchers/active_record/have_rich_text_matcher.rb:
```diff
@@ -0,0 +1,76 @@
+      #
+      # @return [HaveRichText]
+      #
+
```
Can you remove this extra line? I'm not certain that YARD will realize that the above comment block goes with the method.

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+  end
+
+  if rich_text_is_defined?
+    it 'returns the message including the name of the provided attribute' do
```
Mind wrapping this test in a `describe '#description'` so it's clear that's what's being tested instead of the results of the matcher?

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+        to eq('have :content as a ActionText::RichText')
+    end
+
+    context 'a has_rich_text association' do
```
This context is a tad confusing. Usually when we make contexts we are describing a scenario in which something changes. This sounds like we are changing what the subject of the test is. Also, usually whenever you have a context, you have a countercontext. For instance, if you have "when something is the case" then you will also have "when something is not the case". Here you don't have that. So what if you were to unwrap these tests and remove the context?

VSPPedro replied:
Thank you for the lesson!

I refactored the context of the tests!

Tell me what do you think, please.

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+    
+        valid_record = new_post(has_rich_text_association: true)
+    
+        expect(valid_record).to have_rich_text(:content)
```
Right now you have a separate test that tests `failure_message_when_negated`. However, you can also use the `match_against` matcher to do this, just as you've done on line 65. So what if you were to say:

```
expect { have_rich_text(:content) }.
  to match_against(valid_record).
  or_fail_with(expected_message)
```

only here, `expected_message` would be the `failure_message_when_negated`.

VSPPedro replied:
Much better!

Thanks!

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+    include_examples 'undefined rich text', 'String'
+  end
+
+  def create_table_posts(has_content_attribute: false)
```
Can you combine these two methods together? `define_model` calls `create_table` under the hood.

For instance:

``` ruby
def new_post(has_content_attribute: false, is_rich_text_attribute: false)
  columns = {}

  if has_content_attribute
    columns[:content] = :string
  end

  define_model 'Post', columns do
    if has_content_attribute && is_rich_text_attribute
      has_rich_text :content
    end
  end.new
end
```

Also, note that I've renamed the `has_rich_text_association`. We need to be careful when using the word "association". The `has_rich_text` macro does _not_ make `content` an association, so I feel like that could be confusing if we use that word.

VSPPedro replied:
Hi, @mcmire, thank you for your support!

If i’m not wrong has_rich_text macro create a association between the model and the ActionTextRichText as a has_one. e.g.: Post has one ActionText through content.

The problem here was some silly mistakes made by me.

One was the poor description of the tests caused by my rusty English, I need to work on it.

The second was because I used content for the relationship with RichText and used the same name for a attribute. It is easy to think that they are the same...

The content attribute is used for failures tests only. It's for a scenario when the developer try to use a attribute of the model and the matcher raise a warning informing something like: this is not valid, it was expecting a ActionText, but this is a Integer/String/Etc.

I refactored the tests and add your suggestion.

What do you think?

mcmire replied:
You're totally right about the association, I should have checked that. That's what you get for making assumptions, I guess :/ So thanks for clarifying that. I checked over your changes again!

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+    it 'raise a undefined method error when action text is not defined' do
+      create_table_posts
+  
+      invalid_record = -> { new_post(has_rich_text_association: true) }
```
Can we rename `has_rich_text_association` to `is_rich_text_attribute`? Also see my comment on `create_table_posts` together.

mcmire commented on lib/shoulda/matchers/active_record/have_rich_text_matcher.rb:
```diff
@@ -0,0 +1,76 @@
+        end
+
+        def description
+          "have :#{rich_text_attribute} as a ActionText::RichText"
```
This seems a bit clunky from an English perspective. What if we were to rename this to "have configured :#{rich_text_attribute} as a ActionText::RichText attribute"? Then the whole thing reads:

> Expected Post to have configured :content as a ActionText::RichText attribute

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+require 'unit_spec_helper'
+
+describe Shoulda::Matchers::ActiveRecord::HaveRichText, type: :model do
+  shared_examples 'undefined rich text' do |kind|
```
I like that you're trying to keep things DRY in this file, but since each of these shared example groups only have one test in them, I feel like it actually makes things a bit harder to follow when reading over these tests. Would you mind unfolding them? It doesn't feel like a lot of duplication anyway.

VSPPedro replied:
Done!

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+
+describe Shoulda::Matchers::ActiveRecord::HaveRichText, type: :model do
+  shared_examples 'undefined rich text' do |kind|
+    it 'raise a undefined method error when action text is not defined' do
```
"raise" -> "raises"
"action text" -> "ActionText"

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+  end
+
+  shared_examples "invalid attribute" do |kind|
+    it "does not matches when the provided attribute it's not a RichText class" do
```
"matches" -> "match"
"it's" -> "is"
"is not a RichText class" -> "is not a RichText attribute", maybe?

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+        expect(valid_record).to have_rich_text(:content)
+      end
+    
+      it 'does not matches when lacking has_rich_text attribute' do
```
"matches" -> "match"

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,99 @@
+          and_fail_with(expected_message)
+      end
+
+      it 'fails with the appropriate message when negated' do
```
This should be covered already if you use the `match_against` matcher as I've mentioned on line 55.

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,84 @@
+    context 'when the model does not have a RichText association' do
+      it 'does not match when provided with a model attribute' do
+        invalid_record = new_post(has_invalid_content: true)
+        expected_message = 'Expected Post to have configured :invalid_content as a ' \
```
Okay, getting closer. Now that I see this message in its full form, I'm now wondering if it's is as helpful as it could be. For instance, `invalid_content` is defined a string column, but happens to be unset at the time this test is run. Does that information help the user realize what they need to change? I guess what I'm getting at is, would it be enough to say "Expected Post to have configured :invalid_content as a ActionText::RichText association" and not include the type at the end?

mcmire replied:
Hehe 😄 so I might have miscommunicated here. What I meant was that I felt like only _this_ particular message should be changed. The other one that says ":invalid_content does not exist", I think that one makes sense to keep. It's important for people to know that an attribute doesn't exist (we do that for other matches as well), but perhaps not so important to know what the current value of the attribute is. So I think it still makes sense for lines 42 and 63 to say "Expected Post to have configured :invalid_attribute as an ActionText::RichText association, but :invalid_attribute does not exist".

VSPPedro replied:
No you did not.

The problem was, I didn't know how to refactor this:

```
def run_checks
   if !has_attribute?
     ":#{rich_text_attribute} does not exist"
  elsif !has_expected_action_text?
    ":#{rich_text_attribute} is a #{@subject.send(rich_text_attribute).class}"
  end
end
```

The first question that pop up on my mind was: what will be returned by the second condition now? Because it does not need the description in the string. So maybe a empty string? I don't like that.

So in the end I removed the conditional and decided to make all the code more clean, but I you're right again. It's important to know that an attribute doesn't exist. 

I'll try again.

Thanks!

VSPPedro replied:
Done!

What do you think?

mcmire replied:
I like your solution! It's concise and yet understandable.

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,84 @@
+    end
+
+    context 'when the model does not have a RichText association' do
+      it 'does not match when provided with a model attribute' do
```
In both of these tests, we are providing a model attribute to the matcher. It seems like the thing that's different between these two tests is that in one, the attribute being tested exists on the model, whereas in the other, it does not. What if you were to have the names of these tests reflect that?

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+    end
+  else
+    it 'raises a undefined method error when ActionText is not defined' do
```
Sort of a similar thing here — for _both_ of these tests, not just this one, ActionText is not defined (that's why you have them in an `else`). Can you have the names reflect the fact that in one, the attribute being tested exists, whereas in the other it does not?

VSPPedro replied:
Done!

mcmire commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,84 @@
+          or_fail_with(expected_message)
+      end
+
+      it 'returns the message including the name of the provided association' do
```
Do you mind pulling this test out into its own `describe` block so that we know it has to do with `description` and not with running the matcher itself? I'm not sure it needs to be in this `if` block (it should be able to run regardless of whether `ActionText` is available or not). You could put the `describe` before the `if` statement on line 8.

O330oei commented:
asap

mcmire commented:
Everything LGTM now! The only thing left, it looks like, is to fix the merge conflicts. I'm not certain but it might have to do with the generated gemfile.lock file. Would you find looking into that and seeing if you can fix those? Other than that, @guialbuk you have any last thoughts on this?

VSPPedro commented:
Sorry, but I'm not sure what to do.

It's not showing any conflict to me.

`This branch has no conflicts with the base branch`

Regardless of that, I updated my branch with master. Did that work out?

mcmire commented:
@VSPPedro What if you were to try rebasing the branch with master instead of merging master? If you would like to squash you are free to do that as well.

VSPPedro commented:
Done!

guialbuk commented on lib/shoulda/matchers/active_record/have_rich_text_matcher.rb:
```diff
@@ -0,0 +1,79 @@
+      end
+    end
+  end
+end
```
Just some nitpicking: missing am empty end lime here

VSPPedro replied:
Ops, thanks!

Done! :tada:

guialbuk commented on spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb:
```diff
@@ -0,0 +1,85 @@
+      end
+    end.new
+  end
+end
```
And here

mcmire commented:
Looks like CI is green. Merging now!
